node/network/collator-protocol/src/validator_side/mod.rs

```diff
@@ -575,6 +552,76 @@ impl CollationsPerRelayParent {
 	}
 }
 
+/// Any error that can occur when awaiting a collation fetch response.
+#[derive(Debug, thiserror::Error)]
+enum CollationFetchError {
+	#[error("Future was cancelled by the validator side.")]
+	Cancelled,
+	#[error("{0}")]
+	Request(#[from] RequestError),
+}
+
+/// Future that concludes when the collator has responded to our collation fetch request
+/// or the request was cancelled by the validator.
+struct CollationFetchRequest {
+	/// Info about the requested collation.
+	pending_collation: PendingCollation,
+	collator_id: CollatorId,
+	/// Responses from collator.
+	from_collator: BoxFuture<'static, req_res::OutgoingResult<CollationFetchingResponse>>,
+	/// Handle used for checking if this request was cancelled.
+	cancellation_token: CancellationToken,
+	/// A jaeger span corresponding to the lifetime of the request.
+	span: Option<jaeger::Span>,
+	/// A metric histogram for the lifetime of the request
+	_lifetime_timer: Option<metrics::prometheus::prometheus::HistogramTimer>,
+}
+
+impl Future for CollationFetchRequest {
+	type Output = (
+		PendingCollation,
+		CollatorId,
+		std::result::Result<CollationFetchingResponse, CollationFetchError>,
+	);
+
+	fn poll(mut self: Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> Poll<Self::Output> {
+		// First check if this fetch request was cancelled.
+		let cancelled = match std::pin::pin!(self.cancellation_token.cancelled()).poll(cx) {
+			Poll::Ready(()) => true,
+			Poll::Pending => false,
+		};
+
+		if cancelled {
+			self.span.as_mut().map(|s| s.add_string_tag("success", "false"));
+			return Poll::Ready((
+				self.pending_collation.clone(),
+				self.collator_id.clone(),
```

Comment:
nit: these clones are unnecessary and can be dropped with a usage of `Option` or `mem::take`